lib/logstash/filters/dictionary/file.rb

```diff
@@ -0,0 +1,136 @@
+# encoding: utf-8
+require 'rufus-scheduler'
+require "logstash/util/loggable"
+require "logstash/filters/fetch_strategy/file"
+
+java_import 'java.util.concurrent.locks.ReentrantReadWriteLock'
+
+module LogStash module Filters module Dictionary
+  class File
+    def self.create(path, refresh_interval, refresh_behaviour, exact, regex)
+      if /\.y[a]?ml$/.match(path)
+        instance = YamlFile.new(path, refresh_interval, exact, regex)
+      elsif path.end_with?(".json")
+        instance = JsonFile.new(path, refresh_interval, exact, regex)
+      elsif path.end_with?(".csv")
+        instance = CsvFile.new(path, refresh_interval, exact, regex)
+      else
+        raise "Translate: Dictionary #{path} has a non valid format"
+      end
+      if refresh_behaviour == 'merge'
+        instance.set_update_strategy(:merge_dictionary)
+      elsif refresh_behaviour == 'replace'
+        instance.set_update_strategy(:replace_dictionary)
+      else
+        # we really should never get here
+        raise(LogStash::ConfigurationError, "Unknown value for refresh_behaviour=#{refresh_behaviour.to_s}")
+      end
+    end
+
+    include LogStash::Util::Loggable
+    attr_reader :dictionary, :fetch_strategy
+
+    def initialize(path, refresh_interval, exact, regex)
+      @dictionary_path = path
+      @refresh_interval = refresh_interval
+      @short_refresh = @refresh_interval < 300.001
+      rw_lock = java.util.concurrent.locks.ReentrantReadWriteLock.new
+      @write_lock = rw_lock.writeLock
+      @dictionary = Hash.new()
+      @update_method = method(:merge_dictionary)
+      initialize_for_file_type
+      raise_exception = true
+      if exact
+        if regex
+          @fetch_strategy = FetchStrategy::File::ExactRegex.new(@dictionary, rw_lock)
+        else
+          @fetch_strategy = FetchStrategy::File::Exact.new(@dictionary, rw_lock)
+        end
```

Comment:
minor: could be a one-line using `@fetch_strategy = regex ? ... :  ...`

Reply:
I could do that. I was worried about line length readability. As each class has the same constructor signature I could return the class `fetch_strategy_class` and instantiate once.
```
      fetch_strategy_class = if exact
          regex ? FetchStrategy::File::ExactRegex : FetchStrategy::File::Exact
        else
          FetchStrategy::File::RegexUnion
        end
      @fetch_strategy = fetch_strategy_class.new(@dictionary, rw_lock)
```
Clearer? WDYT?

Reply:
meh. I personally prefer the original code but then this is really minor so I leave it up to you.